src/bin/ion/commands/generate/generator.rs

```diff
@@ -501,6 +540,105 @@ impl<'a, L: Language + 'static> CodeGenerator<'a, L> {
+    ///  WrappedScalar {
+    ///     name: vec!["org", "example", "Foo"], // assuming the namespace is `org.example`
+    ///     base_type: FullyQualifiedTypeReference { type_name: vec!["String"], parameters: vec![] }
+    ///     doc_comment: None // There is no doc comment defined in above ISL type def
+    ///     source: IslType {name: "foo", .. } // Represents the `IslType` that is getting converted to `AbstractDataType`
+    ///  }
+    /// )
+    /// ```
+    fn build_wrapped_scalar_from_constraints(
+        &mut self,
+        constraints: &[IslConstraint],
+        code_gen_context: &mut CodeGenContext,
+        parent_isl_type: &IslType,
+    ) -> CodeGenResult<AbstractDataType> {
+        let mut wrapped_scalar_builder = WrappedScalarBuilder::default();
+        wrapped_scalar_builder
+            .name(self.current_type_fully_qualified_name.to_owned())
+            .source(parent_isl_type.to_owned());
+        for constraint in constraints {
+            match constraint.constraint() {
+                IslConstraintValue::Type(isl_type) => {
+                    let type_name = self
+                        .fully_qualified_type_ref_name(isl_type, code_gen_context)?
+                        .ok_or(invalid_abstract_data_type_raw_error(format!(
+                            "Could not determine `FullQualifiedTypeReference` for type {:?}",
+                            isl_type
+                        )))?;
+
+                    // by default fields aren't closed
+                    wrapped_scalar_builder.base_type(type_name);
+                }
+                _ => {
+                    return invalid_abstract_data_type_error(
+                        "Could not determine the abstract data type due to conflicting constraints",
+                    );
+                }
+            }
+        }
+
+        Ok(AbstractDataType::WrappedScalar(
+            wrapped_scalar_builder.build()?,
+        ))
+    }
+
+    /// Builds `AbstractDataType::Scalar` from the given constraints.
+    /// ```
+    /// { type: string }
+    /// ```
+    /// This method builds `AbstractDataTYpe`as following:
+    /// ```
+    /// AbstractDataType::Scalar(
+    ///  Scalar {
+    ///     base_type: FullyQualifiedTypeReference { type_name: vec!["String"], parameters: vec![] }
+    ///     doc_comment: None // There is no doc comment defined in above ISL type def
+    ///     source: IslType { .. } // Represents the `IslType` that is getting converted to `AbstractDataType`
+    ///  }
+    /// )
+    /// ```
+    fn build_scalar_from_constraints(
+        &mut self,
+        constraints: &[IslConstraint],
+        code_gen_context: &mut CodeGenContext,
+        parent_isl_type: &IslType,
+    ) -> CodeGenResult<AbstractDataType> {
+        let mut scalar_builder = ScalarBuilder::default();
+        scalar_builder.source(parent_isl_type.to_owned());
+        for constraint in constraints {
+            match constraint.constraint() {
```

Comment:
There are a couple of issues with the way the constraints are being checked here.

As currently written, this will accept something like `{ type: string, type: int }` without complaint. The presence of a contradiction, like `{ type: string, type: int }` should raise an error because struct ordering isn't guaranteed, which means that the generated output could be unstable. That is an issue that should be resolved now.

On the other hand it will complain about something like `{ type: string, codepoint_length: 20 }`. Even if we don't support generating code for constraints like this, we should not fail because someone wants to include it in their ISL. This doesn't necessarily need to be resolved in this PR as long as we document the limitations here.

(This comment also applies to `build_wrapped_scalar_from_constraints`.)

Reply:
Thats right. I can modify this to add a check for what the base type is for the scalar/wrapped scalar ADT and verify that they don't conflict.

Reply:
currently changed this to only allow a single type constraint in a type definition. Code generator will return an error when another `type` constraint is found.